File: comgongdaeng/DataStructure/BOJ_1918.py
exp = input()
ans = ''
st = []
for c in exp:
    if c >='A' and c<='Z' :
        ans += c
    # 우선 순위 높은 순으로 확인
    # 우선 순위가 같거나 높은 연산자가 스택에 있으면 정답 문자열에 추가
    elif c == '(':
        st.append(c)
    elif c == '*' or c == "/":
        while st and (st[-1]=='*' or st[-1]=='/'):
            ans += st.pop()
        st.append(c)
    elif c == '+' or c == '-':
        while st and st[-1] != '(':
            ans += st.pop()
        st.append(c)
    else: 
        #닫는 괄호인 경우 여는 괄호를 만날 때까지 정답 문자열에 추가
        while st and st[-1] != '(':
            ans += st.pop()
        st.pop()
while st:
    ans += st.pop()
print(ans)

File: naarang/DataStructure/BOJ_1918.js
const fs = require("fs");
let input = fs
  .readFileSync("naarang/DataStructure/input.txt")
  .toString()
  .trim()
  .split("");

let stack = [];
let answer = "";

for (let x of input) {
  // 연산자 우선순위가 같거나 높은 것이 stack의 top에 있으면 pop
  if (x === "+" || x === "-") {
    while (stack.length > 0 && stack[stack.length - 1] !== "(") {
      answer += stack.pop();
    }
    stack.push(x);
  } else if ((x === "*") | (x === "/")) {
    while (
      stack.length > 0 &&
      (stack[stack.length - 1] === "*" || stack[stack.length - 1] === "/")
    ) {
      answer += stack.pop();
    }
    stack.push(x);
  } else if (x === "(") {
    stack.push(x);
  } else if (x === ")") {
    while (stack.length > 0 && stack[stack.length - 1] !== "(") {
      answer += stack.pop();
    }
    // ( 연산자 제거하기
    stack.pop();
  } else {
    answer += x;
  }
}

// 마지막 스택에 남아있는 모든 연잔자 pop
while (stack.length > 0) {
  answer += stack.pop();
}

console.log(answer);
